### PR TITLE
apps: work around issue evicting & destroying old bricks on dead nodes

### LINKED ISSUE #1756: Error Removing Dead Node
### Kind of issue
Unsure - perhaps a Bug? (This is a continuation of #1744)

### Observed behavior

We're seeing a scenario where within a 3 node glusterfs cluster one node has become corrupt/unreachable. At which point we have provisioned a new node (to temporarily make a 4 node gluster cluster) and are trying to remove the corrupt node in order to migrate the bricks to the new node to acheive a 3-replica cluster again (currently all volumes have only 2 bricks). This is failing with the following error indicating that it's trying to SSH to the dead node (IP Address = 192.168.3.25) in order to remove the bricks that are being deleted, the issue is that this would never work as the node is down and won't be back up, so we need it to skip this step and simply create the brick on the new 4th node instead.

heketi-cli node remove 0ac9ed991748d6eeb8659db6b40e7094

```
[cmdexec] WARNING 2020/07/27 09:57:12 Failed to create SSH connection to 192.168.3.25:22: dial tcp 192.168.3.25:22: connect: no route to host
[heketi] ERROR 2020/07/27 09:57:12 heketi/apps/glusterfs/operations_device.go:661:glusterfs.(*BrickEvictOperation).Clean: Unable to get volume info from gluster node 192.168.3.25 for volume vol_0371ee52c589f64b27fa942a289d17b8: Unable to get volume info of volume name: vol_0371ee52c589f64b27fa942a289d17b8
[heketi] ERROR 2020/07/27 09:57:12 heketi/apps/glusterfs/operations_manage.go:270:glusterfs.rollbackViaClean: error running Clean in rollback for Remove Brick from Device: Unable to get volume info of volume name: vol_0371ee52c589f64b27fa942a289d17b8
[heketi] ERROR 2020/07/27 09:57:12 heketi/apps/glusterfs/operations_manage.go:175:glusterfs.runOperationAfterBuild: Remove Brick from Device Rollback error: Unable to get volume info of volume name: vol_0371ee52c589f64b27fa942a289d17b8
[heketi] ERROR 2020/07/27 09:57:12 heketi/apps/glusterfs/operations_manage.go:251:glusterfs.RunOperation.func1: Error in Remove Brick from Device: Error destroying old brick: dial tcp 192.168.3.25:22: connect: no route to host
[heketi] ERROR 2020/07/27 09:57:12 heketi/apps/glusterfs/operations_manage.go:167:glusterfs.runOperationAfterBuild: Remove Device Failed: Error destroying old brick: dial tcp 192.168.3.25:22: connect: no route to host
[heketi] INFO 2020/07/27 09:57:12 need to clean child [1cafcc6644680acd8c4f37ce6e719283] of Remove Device [9ad79dd9f9cd802ef2fffc62ce93a78b]
[heketi] INFO 2020/07/27 09:57:12 Starting Clean for Evict Brick op:1cafcc6644680acd8c4f37ce6e719283
[heketi] INFO 2020/07/27 09:57:12 found brick to be replaced: 1836f4f73ffd968b68e00fe04129b470
[heketi] INFO 2020/07/27 09:57:12 this is a child op of: 9ad79dd9f9cd802ef2fffc62ce93a78b
[heketi] INFO 2020/07/27 09:57:12 found replacement brick: 727e9b05d75eaac0c264380cdbd1f516
[negroni] 2020-07-27T09:57:12Z | 200 |   133.782µs | localhost:8080 | GET /queue/a9a244c0ca8c2bfe357409f6f754393f
[negroni] 2020-07-27T09:57:12Z | 200 |   172.284µs | localhost:8080 | GET /queue/a9a244c0ca8c2bfe357409f6f754393f
[negroni] 2020-07-27T09:57:13Z | 200 |   83.151µs | localhost:8080 | GET /queue/a9a244c0ca8c2bfe357409f6f754393f
[negroni] 2020-07-27T09:57:13Z | 200 |   146.811µs | localhost:8080 | GET /queue/a9a244c0ca8c2bfe357409f6f754393f
[negroni] 2020-07-27T09:57:13Z | 200 |   139.907µs | localhost:8080 | GET /queue/a9a244c0ca8c2bfe357409f6f754393f
[negroni] 2020-07-27T09:57:14Z | 200 |   126.986µs | localhost:8080 | GET /queue/a9a244c0ca8c2bfe357409f6f754393f
[negroni] 2020-07-27T09:57:14Z | 200 |   193.893µs | localhost:8080 | GET /queue/a9a244c0ca8c2bfe357409f6f754393f
[negroni] 2020-07-27T09:57:15Z | 200 |   146.202µs | localhost:8080 | GET /queue/a9a244c0ca8c2bfe357409f6f754393f
[cmdexec] WARNING 2020/07/27 09:57:15 Failed to create SSH connection to 192.168.3.25:22: dial tcp 192.168.3.25:22: connect: no route to host
[heketi] ERROR 2020/07/27 09:57:15 heketi/apps/glusterfs/operations_device.go:661:glusterfs.(*BrickEvictOperation).Clean: Unable to get volume info from gluster node 192.168.3.25 for volume vol_0371ee52c589f64b27fa942a289d17b8: Unable to get volume info of volume name: vol_0371ee52c589f64b27fa942a289d17b8
[heketi] ERROR 2020/07/27 09:57:15 heketi/apps/glusterfs/operations_manage.go:270:glusterfs.rollbackViaClean: error running Clean in rollback for Remove Device: Unable to get volume info of volume name: vol_0371ee52c589f64b27fa942a289d17b8
[heketi] ERROR 2020/07/27 09:57:15 heketi/apps/glusterfs/operations_manage.go:175:glusterfs.runOperationAfterBuild: Remove Device Rollback error: Unable to get volume info of $olume name: vol_0371ee52c589f64b27fa942a289d17b8
```

### Expected/desired behavior

It shouldn't bother deleting the bricks on a node that is being removed as it could be down. In our case as it's corrupt/inaccessible we can't migrate the bricks to the new node/remove the node.

### Details on how to reproduce (minimal and precise)

As with the previous issue (#1744), we have tried to replicate this issue in another heketi/gluster instance but it seemed to work fine - we shut down one node in the cluster and did a node delete to force migration of the bricks onto another node which worked without issue. I don't know if this behavior is something that has been introduced in recent releases or is expected.

### Information about the environment:

- Heketi version used  (e.g. v6.0.0 or master): v9.0.0-204-gd23f5303
- Operating system used: Gluster Cluster - Ubuntu 18.04, Heketi via Docker image
- Heketi compiled from sources, as a package (rpm/deb), or container: Container
- If container, which container image: heketi/heketi:dev 
- Using kubernetes, openshift, or direct install: Kubernetes
- If kubernetes/openshift, is gluster running inside kubernetes/openshift or outside: Outside
- If kubernetes/openshift, how was it deployed (gk-deploy, openshift-ansible, other, custom): gk-deploy 

Again any help pointing in the right direction would be very helpful.

### PR DESCRIPTION
**Pull request by phlogistonjohn**
### What does this PR achieve? Why do we need it?

If a brick destroy calls fails during brick evict and the node is
down, do not treat it as a fatal error. If the node or device replace
is being done for a permanently down node then the destroy call will
never succeed. Use the ability of heketi to probe the node to decide
if if the error is fatal (node responsive) or not (node unresponsive).

There's one flaw with this: if the node is temporarily down. In that
case we could be orphaning bricks on said node, and when it comes up
that node would be out of sync with heketi. A future direction might
be to add a new node state to indicate "offline-and-never-coming-back"
that indicate that errors in this code path are truly safe to ignore.

### Does this PR fix issues?

Fixes #1756  (maybe, needs validation)

### Notes for the reviewer

Untested, needs practical validation. But I'm throwing this out as a conversation starter.

### CODE REVIEW

**Comment by iamniting** [comment]
Hi @phlogistonjohn 

I have tested the PR before and after pulling your changes to the master branch in my local cluster. 

Tests I performed:
 - created 4 node cluster with 4 zones
 - disabled the 4th node
 - created 5 replicas 3 volumes. Which is divided across the first 3 nodes
 - powered off the first node
 - evicted a brick, got an error as changes were not applied yet
```
# heketi-cli brick evict 1ca81c4f414f6d1760ac0fa813de3d6f
Error: Error destroying old brick: dial tcp 10.70.53.153:22: connect: no route to host
```
 - brick evict went into pending operations
```
# heketi-cli server operations list
Id:a763c99d59728159dc124f78ad9e2404  Type:evict-brick  Status:stale
```
- applied your changes
- again ran the brick evict, it got failed as it was in pending operations
```
# heketi-cli brick evict 1ca81c4f414f6d1760ac0fa813de3d6f
Error: Failed to set up brick eviction: Can not evict brick 1ca81c4f414f6d1760ac0fa813de3d6f: brick is pending
```
 - ran brick evict on other brick it was successful and added the brick into 4th node as expected
```
# heketi-cli brick evict 3143bf680cadfb8f11e2c6ad6d1ce3ad
Brick 3143bf680cadfb8f11e2c6ad6d1ce3ad evicted
```
 - but I am not able to evict the older brick which was failing, I tried operation cleanup. I am seeing the same error in logs as defined in issue. I want to understand more about how pending operations work so I can try to solve it

logs:

```
[asynchttp] INFO 2020/07/28 12:05:59 Started job bdc5ca1904c8f0aba22d35d26dc16ce1
[heketi] DEBUG 2020/07/28 12:05:59 heketi/apps/glusterfs/operations_cleanup.go:33:glusterfs.OperationCleaner.Clean: Going to clean up operations
[heketi] DEBUG 2020/07/28 12:05:59 heketi/apps/glusterfs/operations_cleanup.go:231:glusterfs.CleanSelectedOps.func1: matched pending operation id: a763c99d59728159dc124f78ad9e2404
[heketi] INFO 2020/07/28 12:05:59 Found operation a763c99d59728159dc124f78ad9e2404 in need of clean up
[heketi] INFO 2020/07/28 12:05:59 Starting Clean for Evict Brick op:a763c99d59728159dc124f78ad9e2404
[heketi] INFO 2020/07/28 12:05:59 found brick to be replaced: 1ca81c4f414f6d1760ac0fa813de3d6f
[heketi] INFO 2020/07/28 12:05:59 found replacement brick: 864640dbe79ff661964175e84976b486
[negroni] 2020-07-28T12:05:59+05:30 | 200 | 	 96.332µs | localhost:8080 | GET /queue/bdc5ca1904c8f0aba22d35d26dc16ce1
[negroni] 2020-07-28T12:06:00+05:30 | 200 | 	 113.921µs | localhost:8080 | GET /queue/bdc5ca1904c8f0aba22d35d26dc16ce1
[negroni] 2020-07-28T12:06:00+05:30 | 200 | 	 107.754µs | localhost:8080 | GET /queue/bdc5ca1904c8f0aba22d35d26dc16ce1
[negroni] 2020-07-28T12:06:00+05:30 | 200 | 	 94.476µs | localhost:8080 | GET /queue/bdc5ca1904c8f0aba22d35d26dc16ce1
[negroni] 2020-07-28T12:06:01+05:30 | 200 | 	 402.606µs | localhost:8080 | GET /queue/bdc5ca1904c8f0aba22d35d26dc16ce1
[negroni] 2020-07-28T12:06:01+05:30 | 200 | 	 215.087µs | localhost:8080 | GET /queue/bdc5ca1904c8f0aba22d35d26dc16ce1
[negroni] 2020-07-28T12:06:02+05:30 | 200 | 	 205.671µs | localhost:8080 | GET /queue/bdc5ca1904c8f0aba22d35d26dc16ce1
[negroni] 2020-07-28T12:06:02+05:30 | 200 | 	 207.06µs | localhost:8080 | GET /queue/bdc5ca1904c8f0aba22d35d26dc16ce1
[cmdexec] WARNING 2020/07/28 12:06:02 Failed to create SSH connection to dhcp53-153.lab.eng.blr.redhat.com:22: dial tcp 10.70.53.153:22: connect: no route to host
[heketi] WARNING 2020/07/28 12:06:02 Clean phase of operation a763c99d59728159dc124f78ad9e2404 encountered error: Unable to get volume info of volume name: vol2
[heketi] WARNING 2020/07/28 12:06:02 Unable to clean operation a763c99d59728159dc124f78ad9e2404: Unable to get volume info of volume name: vol2
[asynchttp] INFO 2020/07/28 12:06:02 Completed job bdc5ca1904c8f0aba22d35d26dc16ce1 in 3.102913703s
[negroni] 2020-07-28T12:06:02+05:30 | 204 | 	 110.094µs | localhost:8080 | GET /queue/bdc5ca1904c8f0aba22d35d26dc16ce1
```

**Comment by phlogistonjohn** [comment]
We did a bunch of practical tests and found some more issues, new patches have been pushed.

File: apps/glusterfs/operations_device.go [hunk 1]
```diff
@@ -592,7 +592,21 @@ func (beo *BrickEvictOperation) execReplaceBrick(
 
 	beo.reclaimed, err = old.bhmap().destroy(executor)
 	if err != nil {
-		return logger.LogError("Error destroying old brick: %v", err)
+		// If the destroy failed because the node is not running (ideally,
+		// permanently down) we want to continue with the operation . But if
+		// the node works we failed for a "real" reason we don't want to
+		// blindly ignore errors. So we do a probe to see if the host is
+		// responsive. If the destroy has failed we will use the node's
+		// responsiveness to decide if // this should be treated as a failure
```
Comment (iamniting) [inline apps/glusterfs/operations_device.go]: **//** can be removed after if statement.

File: apps/glusterfs/operations_device.go [hunk 2]
```diff
@@ -1033,3 +1016,29 @@ func getWorkingNode(n *NodeEntry, db wdb.RODB, executor executors.Executor) (str
...
+	if err == nil {
+		return reclaimed, nil
+	}
+
+	// If the destroy failed because the node is not running (ideally,
+	// permanently down) we want to continue with the operation . But if
+	// the node works we failed for a "real" reason we don't want to
+	// blindly ignore errors. So we do a probe to see if the host is
+	// responsive. If the destroy has failed we will use the node's
+	// responsiveness to decide if // this should be treated as a failure
```
Comment (iamniting) [inline apps/glusterfs/operations_device.go:1031]: here also **//** can be removed in b/w the line

**Comment by iamniting** [comment]
@phlogistonjohn all changes looks good to me. There are minor changes in comments. Which is good to have. But I would not insist on that as they are comments. I leave it up to you.

File: apps/glusterfs/operations_device.go (hunk 2 above)
Comment (phlogistonjohn) [reply apps/glusterfs/operations_device.go]: Fixed. Thanks.

**Comment by phlogistonjohn** [comment]
Centos CI passed, and all the relevant tests in the travis ci job worked (python tests failed due to some environmental issue). WIth that an 2 reviews I think we're good to merge this. @raghavendra-talur care to do the honors? I feel like I'm cheating a bit if I force merge my own patches.

**Pull request merged**